File: packages/typescript/src/Daytona.ts
/**
 * Sandboxes are isolated development environments managed by Daytona.
 * This guide covers how to create, manage, and remove Sandboxes using the SDK.
 * 
 * @module Daytona
 * 
 * @example
 * // Initialize using environment variables (DAYTONA_API_KEY, DAYTONA_SERVER_URL, DAYTONA_TARGET)
 * const daytona = new Daytona();
 * 
 * // Create and use a sandbox
 * const sandbox = await daytona.create({
 *     language: 'typescript',
 *     envVars: { NODE_ENV: 'development' }
 * });
 * 
 * // Execute commands in the sandbox
 * const response = await sandbox.process.executeCommand('echo "Hello, World!"');
 * console.log(response.result);
 * 
 * // Execute code in the sandbox
 * const response = await sandbox.process.codeRun('console.log("Hello, World!")');
 * console.log(response.result);
 * 
 * @example
 * // Initialize with explicit configuration
 * const daytona = new Daytona({
 *     apiKey: process.env.CUSTOM_API_KEY,
 *     serverUrl: 'https://daytona.example.com',
 *     target: 'us'
 * });
 * 
 * // Create a custom sandbox
 * const sandbox = await daytona.create({
 *     language: 'typescript',
 *     image: 'node:18',
 *     resources: {
 *         cpu: 2,
 *         memory: 4 // 4GB RAM
 *     },
 *     autoStopInterval: 60 // Auto-stop after 1 hour of inactivity
 * });
 * 
 * // Use sandbox features
 * await sandbox.git.clone('https://github.com/user/repo.git');
 * await sandbox.process.executeCommand('npm test');
 */

import { SandboxPythonCodeToolbox } from './code-toolbox/SandboxPythonCodeToolbox'
import { Sandbox, SandboxInstance, Sandbox as Workspace } from './Sandbox'
import {
  Configuration,
  WorkspaceApi as SandboxApi,
  ToolboxApi,
  CreateWorkspaceTargetEnum as SandboxTargetRegion
} from '@daytonaio/api-client'
import { SandboxTsCodeToolbox } from './code-toolbox/SandboxTsCodeToolbox'
import axios, { AxiosError } from 'axios'
import { DaytonaError } from './errors/DaytonaError'

/**
 * Configuration options for initializing the Daytona client.
 * 
 * @interface
 * @property {string} apiKey - API key for authentication with Daytona server
 * @property {string} serverUrl - URL of the Daytona server
 * @property {CreateSandboxTargetEnum} target - Target location for Sandboxes
 * 
 * @example
 * const config: DaytonaConfig = {
 *     apiKey: "your-api-key",
 *     serverUrl: "https://your-server.com",
 *     target: "us"
 * };
 * const daytona = new Daytona(config);
 */
export interface DaytonaConfig {
  /** API key for authentication with Daytona server */
  apiKey: string
  /** URL of the Daytona server */
  serverUrl: string
  /** Target environment for sandboxes */
  target: SandboxTargetRegion
}

/**
 * Supported programming languages for code execution
 */
export enum CodeLanguage {
    PYTHON = "python",
    TYPESCRIPT = "typescript",
    JAVASCRIPT = "javascript",
}

/**
 * Resource allocation for a Sandbox.
 * 
 * @interface
 * @property {number} [cpu] - CPU allocation for the Sandbox in cores
 * @property {number} [gpu] - GPU allocation for the Sandbox in units
 * @property {number} [memory] - Memory allocation for the Sandbox in GB
 * @property {number} [disk] - Disk space allocation for the Sandbox in GB
 * 
 * @example
 * const resources: SandboxResources = {
 *     cpu: 2,
 *     memory: 4,  // 4GB RAM
 *     disk: 20    // 20GB disk
 * };
 */
export interface SandboxResources {
  /** CPU allocation for the Sandbox */
  cpu?: number
  /** GPU allocation for the Sandbox */
  gpu?: number
  /** Memory allocation for the Sandbox in GB */
  memory?: number
  /** Disk space allocation for the Sandbox in GB */
  disk?: number
}

/**
 * Parameters for creating a new Sandbox.
 * 
 * @interface
 * @property {string} [id] - Optional Sandbox ID. If not provided, a random ID will be generated
 * @property {string} [image] - Optional Docker image to use for the Sandbox
 * @property {string} [user] - Optional os user to use for the Sandbox
 * @property {CodeLanguage} [language] - Programming language for direct code execution
 * @property {Record<string, string>} [envVars] - Optional environment variables to set in the Sandbox
 * @property {Record<string, string>} [labels] - Sandbox labels
 * @property {boolean} [public] - Is the Sandbox port preview public
 * @property {string} [target] - Target location for the Sandbox
 * @property {SandboxResources} [resources] - Resource allocation for the Sandbox
 * @property {boolean} [async] - If true, will not wait for the Sandbox to be ready before returning
 * @property {number} [timeout] - Timeout in seconds for the Sandbox to be ready (0 means no timeout)
 * @property {number} [autoStopInterval] - Auto-stop interval in minutes (0 means disabled)
 * 
 * @example
 * const params: CreateSandboxParams = {
 *     language: 'typescript',
 *     envVars: { NODE_ENV: 'development' },
 *     resources: {
 *         cpu: 2,
 *         memory: 4 // 4GB RAM
 *     },
 *     autoStopInterval: 60  // Auto-stop after 1 hour of inactivity
 * };
 * const sandbox = await daytona.create(params, 50);
 */
export type CreateSandboxParams = {
  /** Optional Sandbox ID. If not provided, a random ID will be generated */
  id?: string
  /** Optional Docker image to use for the Sandbox */
  image?: string
  /** Optional os user to use for the Sandbox */
  user?: string
  /** Programming language for direct code execution */
  language?: CodeLanguage | string
  /** Optional environment variables to set in the sandbox */
  envVars?: Record<string, string>
  /** Sandbox labels */
  labels?: Record<string, string>
  /** Is the Sandbox port preview public */
  public?: boolean
  /** Target location for the Sandbox */
  target?: SandboxTargetRegion | string
  /** Resource allocation for the Sandbox */
  resources?: SandboxResources
  /** If true, will not wait for the Sandbox to be ready before returning */
  async?: boolean
  /**
   * Timeout in seconds, for the Sandbox to be ready (0 means no timeout)
   * @deprecated Use methods with `timeout` parameter instead
   */
  timeout?: number
  /** Auto-stop interval in minutes (0 means disabled) (must be a non-negative integer) */
  autoStopInterval?: number
}

/**
 * Main class for interacting with Daytona Server API.
 * 
 * Provides methods for creating, managing, and interacting with Daytona Sandboxes.
 * Can be initialized either with explicit configuration or using environment variables.
 * 
 * 
 * @example 
 * // Using environment variables
 * // Uses DAYTONA_API_KEY, DAYTONA_SERVER_URL, DAYTONA_TARGET
 * const daytona = new Daytona();
 * const sandbox = await daytona.create();
 * 
 * @example 
 * // Using explicit configuration
 * const config: DaytonaConfig = {
 *     apiKey: "your-api-key",
 *     serverUrl: "https://your-server.com",
 *     target: "us"
 * };
 * const daytona = new Daytona(config);
 * 
 * @class
 */
export class Daytona {
  private readonly sandboxApi: SandboxApi
  private readonly toolboxApi: ToolboxApi
  private readonly target: SandboxTargetRegion
  private readonly apiKey: string
  private readonly serverUrl: string

  /**
   * Creates a new Daytona client instance.
   * 
   * @param {DaytonaConfig} [config] - Configuration options
   * @throws {DaytonaError} - `DaytonaError` - When API key or server URL is missing
   */
  constructor(config?: DaytonaConfig) {
    const apiKey = config?.apiKey || process.env.DAYTONA_API_KEY
    if (!apiKey) {
      throw new DaytonaError('API key is required')
    }
    const serverUrl = config?.serverUrl || process.env.DAYTONA_SERVER_URL || 'https://app.daytona.io/api'
    const envTarget = process.env.DAYTONA_TARGET as SandboxTargetRegion
    const target = config?.target || envTarget || SandboxTargetRegion.US

    this.apiKey = apiKey
    this.serverUrl = serverUrl
    this.target = target

    const configuration = new Configuration({
      basePath: this.serverUrl,
      baseOptions: {
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
        },
      },
    })

    const axiosInstance = axios.create();
    axiosInstance.interceptors.response.use(
      (response) => {
        return response
      },
      (error) => {
        let errorMessage: string;

        if (error instanceof AxiosError && error.message.includes('timeout of')) {
          errorMessage = "Operation timed out"
        } else {
          errorMessage =
            error.response?.data?.message ||
            error.response?.data ||
            error.message ||
            String(error);
        }

        throw new DaytonaError(errorMessage);
      }
    )

    this.sandboxApi = new SandboxApi(configuration, '', axiosInstance)
    this.toolboxApi = new ToolboxApi(configuration, '', axiosInstance)
  }

  /**
   * Creates Sandboxes with default or custom configurations. You can specify various parameters,
   * including language, image, resources, environment variables, and volumes for the Sandbox.
   * 
   * @param {CreateSandboxParams} [params] - Parameters for Sandbox creation
   * @param {number} [timeout] - Timeout in seconds (0 means no timeout, default is 60)
   * @returns {Promise<Sandbox>} The created Sandbox instance
   * 
   * @example
   * // Create a default sandbox
   * const sandbox = await daytona.create();
   * 
   * @example
   * // Create a custom sandbox
   * const params: CreateSandboxParams = {
   *     language: 'typescript',
   *     image: 'node:18',
   *     envVars: { 
   *         NODE_ENV: 'development',
   *         DEBUG: 'true'
   *     },
   *     resources: {
   *         cpu: 2,
   *         memory: 4 // 4GB RAM
   *     },
   *     autoStopInterval: 60
   * };
   * const sandbox = await daytona.create(params, 40);
   */
  public async create(params?: CreateSandboxParams, timeout: number = 60): Promise<Sandbox> {
    // const startTime = Date.now();

    if (params == null) {
      params = { language: 'python' }
    }

    const labels = params.labels || {}
    if (params.language) {
      labels[`code-toolbox-language`] = params.language
    }

    // remove this when params.timeout is removed
    const effectiveTimeout = params.timeout || timeout
    if (effectiveTimeout < 0) {
      throw new DaytonaError('Timeout must be a non-negative number')
    }

    if (params.autoStopInterval !== undefined && (!Number.isInteger(params.autoStopInterval) || params.autoStopInterval < 0)) {
      throw new DaytonaError('autoStopInterval must be a non-negative integer');
    }

    const codeToolbox = this.getCodeToolbox(params.language as CodeLanguage)

    try {
      const response = await this.sandboxApi.createWorkspace({
          id: params.id,
          name: params.id,
          image: params.image,
          user: params.user,
          env: params.envVars || {},
          labels: params.labels,
          public: params.public,
          target: this.target,
          cpu: params.resources?.cpu,
          gpu: params.resources?.gpu,
          memory: params.resources?.memory,
          disk: params.resources?.disk,
          autoStopInterval: params.autoStopInterval,
        },
        {
          timeout: effectiveTimeout * 1000
        })

      const sandboxInstance = response.data
      const sandboxInfo = Sandbox.toSandboxInfo(sandboxInstance)
      sandboxInstance.info = sandboxInfo

      const sandbox = new Sandbox(
        sandboxInstance.id!,
        sandboxInstance as SandboxInstance,
        this.sandboxApi,
        this.toolboxApi,
        codeToolbox,
      )

      // if (!params.async) {
      //   const timeElapsed = Date.now() - startTime;
      //   await sandbox.waitUntilStarted(effectiveTimeout ? effectiveTimeout - (timeElapsed / 1000) : 0);
      // }

      return sandbox
    } catch (error) {
      void this.sandboxApi.deleteWorkspace(params.id!, true).catch(() => {});
      if (error instanceof DaytonaError && error.message.includes("Operation timed out")) {
        throw new DaytonaError(`Failed to create and start sandbox within ${effectiveTimeout} seconds. Operation timed out.`)
      }
      throw error
    }
  }

  /**
   * Gets a Sandbox by its ID.
   * 
   * @param {string} sandboxId - The ID of the Sandbox to retrieve
   * @returns {Promise<Sandbox>} The Sandbox
   * 
   * @example
   * const sandbox = await daytona.get('my-sandbox-id');
   * console.log(`Sandbox state: ${sandbox.instance.state}`);
   */
  public async get(sandboxId: string): Promise<Sandbox> {
    const response = await this.sandboxApi.getWorkspace(sandboxId)
    const sandboxInstance = response.data
    const language = sandboxInstance.labels && sandboxInstance.labels[`code-toolbox-language`]
    const codeToolbox = this.getCodeToolbox(language as CodeLanguage)
    const sandboxInfo = Sandbox.toSandboxInfo(sandboxInstance)
    sandboxInstance.info = sandboxInfo

    return new Sandbox(sandboxId, sandboxInstance as SandboxInstance, this.sandboxApi, this.toolboxApi, codeToolbox)
  }

  /**
   * Lists all Sandboxes.
   * 
   * @returns {Promise<Sandbox[]>} Array of Sandboxes
   * 
   * @example
   * const sandboxes = await daytona.list();
   * for (const sandbox of sandboxes) {
   *     console.log(`${sandbox.id}: ${sandbox.instance.state}`);
   * }
   */
  public async list(): Promise<Sandbox[]> {
    const response = await this.sandboxApi.listWorkspaces()
    return response.data.map((sandbox) => {
      const language = sandbox.labels?.[`code-toolbox-language`] as CodeLanguage
      const sandboxInfo = Sandbox.toSandboxInfo(sandbox)
      sandbox.info = sandboxInfo

      return new Sandbox(
        sandbox.id, 
        sandbox as SandboxInstance, 
        this.sandboxApi, 
        this.toolboxApi, 
        this.getCodeToolbox(language)
      )
    })
  }

  /**
   * Starts a Sandbox and waits for it to be ready.
   * 
   * @param {Sandbox} sandbox - The Sandbox to start
   * @param {number} [timeout] - Optional timeout in seconds (0 means no timeout)
   * @returns {Promise<void>}
   * 
   * @example
   * const sandbox = await daytona.get('my-sandbox-id');
   * // Wait up to 60 seconds for the sandbox to start
   * await daytona.start(sandbox, 60);
   */
  public async start(sandbox: Sandbox, timeout?: number) {
    await sandbox.start(timeout)
  }

  /**
   * Stops a Sandbox.
   * 
   * @param {Sandbox} sandbox - The Sandbox to stop
   * @returns {Promise<void>}
   * 
   * @example
   * const sandbox = await daytona.get('my-sandbox-id');
   * await daytona.stop(sandbox);
   */
  public async stop(sandbox: Sandbox) {
    await sandbox.stop()
  }

  /**
   * Removes a Sandbox.
   * 
   * @param {Sandbox} sandbox - The Sandbox to remove
   * @param {number} timeout - Timeout in seconds (0 means no timeout, default is 60)
   * @returns {Promise<void>}
   * 
   * @example
   * const sandbox = await daytona.get('my-sandbox-id');
   * await daytona.remove(sandbox);
   */
  public async remove(sandbox: Sandbox, timeout: number = 60) {
    await this.sandboxApi.deleteWorkspace(sandbox.id, true, { timeout: timeout * 1000 })
  }

  /**
   * Gets the Sandbox by ID.
   * 
   * @param {string} workspaceId - The ID of the Sandbox to retrieve
   * @returns {Promise<Workspace>} The Sandbox
   * 
   * @deprecated Use `getCurrentSandbox` instead. This method will be removed in a future version.
   */
  public async getCurrentWorkspace(workspaceId: string): Promise<Workspace> {
    return await this.getCurrentSandbox(workspaceId)
  }

  /**
   * Gets the Sandbox by ID.
   * 
   * @param {string} sandboxId - The ID of the Sandbox to retrieve
   * @returns {Promise<Sandbox>} The Sandbox
   * 
   * @example
   * const sandbox = await daytona.getCurrentSandbox('my-sandbox-id');
   * console.log(`Current sandbox state: ${sandbox.instance.state}`);
   */
  public async getCurrentSandbox(sandboxId: string): Promise<Sandbox> {
    return await this.get(sandboxId)
  }

  /**
   * Gets the appropriate code toolbox based on language.
   * 
   * @private
   * @param {CodeLanguage} [language] - Programming language for the toolbox
   * @returns {SandboxCodeToolbox} The appropriate code toolbox instance
   * @throws {DaytonaError} - `DaytonaError` - When an unsupported language is specified
   */
  private getCodeToolbox(language?: CodeLanguage) {
    switch (language) {
      case CodeLanguage.JAVASCRIPT:
      case CodeLanguage.TYPESCRIPT:
        return new SandboxTsCodeToolbox()
      case CodeLanguage.PYTHON:
      case undefined:
        return new SandboxPythonCodeToolbox()
      default:
        throw new DaytonaError(`Unsupported language: ${language}, supported languages: ${Object.values(CodeLanguage).join(', ')}`)
    }
  }
}

File: packages/python/src/daytona_sdk/daytona.py
"""
Provides the main entry point for interacting with Daytona Server API.

The Daytona allows you to create, manage, and interact with Daytona workspaces.
It provides a high-level interface for workspace operations including creation,
deletion, listing, and status management.

Example:
    Basic usage with environment variables:
    ```python
    from daytona_sdk import Daytona
    daytona = Daytona()  # Uses env vars DAYTONA_API_KEY, DAYTONA_SERVER_URL
    workspace = daytona.create()  # Creates a Python workspace
    # Run the code securely inside the workspace
    response = workspace.process.code_run('print("Hello World!")')
    print(response.result)
    daytona.remove(workspace)
    ```

    Usage with explicit configuration:
    ```python
    config = DaytonaConfig(
        api_key="your-api-key",
        server_url="https://your-server.com",
        target="us"
    )
    daytona = Daytona(config)
    ```
"""

import uuid
import json
from typing import Optional, Literal, Dict, Any, List
from dataclasses import dataclass
from environs import Env
import time
from daytona_api_client import (
    Configuration,
    WorkspaceApi,
    ToolboxApi,
    ApiClient,
    CreateWorkspace,
    SessionExecuteRequest,
    SessionExecuteResponse
)

from .code_toolbox.workspace_python_code_toolbox import WorkspacePythonCodeToolbox
from .code_toolbox.workspace_ts_code_toolbox import WorkspaceTsCodeToolbox
from .workspace import Workspace


# Type definitions
CodeLanguage = Literal["python", "javascript", "typescript"]


@dataclass
class DaytonaConfig:
    """Configuration options for initializing the Daytona client.

    Attributes:
        api_key (str): API key for authentication with Daytona server.
        server_url (str): URL of the Daytona server.
        target (str): Target environment for workspaces.

    Example:
        ```python
        config = DaytonaConfig(
            api_key="your-api-key",
            server_url="https://your-server.com",
            target="us"
        )
        daytona = Daytona(config)
        ```
    """
    api_key: str
    server_url: str
    target: str


@dataclass
class WorkspaceResources:
    """Resources configuration for workspace.

    Attributes:
        cpu (Optional[int]): Number of CPU cores to allocate.
        memory (Optional[int]): Amount of memory in GB to allocate.
        disk (Optional[int]): Amount of disk space in GB to allocate.
        gpu (Optional[int]): Number of GPUs to allocate.

    Example:
        ```python
        resources = WorkspaceResources(
            cpu=2,
            memory=4,  # 4GB RAM
            disk=20,   # 20GB disk
            gpu=1
        )
        params = CreateWorkspaceParams(
            language="python",
            resources=resources
        )
        ```
    """
    cpu: Optional[int] = None
    memory: Optional[int] = None
    disk: Optional[int] = None
    gpu: Optional[int] = None


@dataclass
class CreateWorkspaceParams:
    """Parameters for creating a new workspace.

    Attributes:
        language (CodeLanguage): Programming language for the workspace ("python", "javascript", "typescript").
        id (Optional[str]): Custom identifier for the workspace. If not provided, a random ID will be generated.
        name (Optional[str]): Display name for the workspace. Defaults to workspace ID if not provided.
        image (Optional[str]): Custom Docker image to use for the workspace.
        os_user (Optional[str]): OS user for the workspace. Defaults to "daytona".
        env_vars (Optional[Dict[str, str]]): Environment variables to set in the workspace.
        labels (Optional[Dict[str, str]]): Custom labels for the workspace.
        public (Optional[bool]): Whether the workspace should be public.
        target (Optional[str]): Target location for the workspace. Can be "us", "eu", or "asia".
        resources (Optional[WorkspaceResources]): Resource configuration for the workspace.
        timeout (Optional[float]): Timeout in seconds for workspace to be created and started.
        auto_stop_interval (Optional[int]): Interval in minutes after which workspace will automatically stop if no workspace event occurs during that time. Default is 15 minutes. If set to 0, the workspace will not be automatically stopped.

    Example:
        ```python
        params = CreateWorkspaceParams(
            language="python",
            name="my-workspace",
            env_vars={"DEBUG": "true"},
            resources=WorkspaceResources(cpu=2, memory=4),
            timeout=60,
            auto_stop_interval=20
        )
        workspace = daytona.create(params)
        ```
    """
    language: CodeLanguage
    id: Optional[str] = None
    name: Optional[str] = None
    image: Optional[str] = None
    os_user: Optional[str] = None
    env_vars: Optional[Dict[str, str]] = None
    labels: Optional[Dict[str, str]] = None
    public: Optional[bool] = None
    target: Optional[str] = None
    resources: Optional[WorkspaceResources] = None
    timeout: Optional[float] = None
    auto_stop_interval: Optional[int] = None


class Daytona:
    """Main class for interacting with Daytona Server API.

    This class provides methods to create, manage, and interact with Daytona workspaces.
    It can be initialized either with explicit configuration or using environment variables.

    Attributes:
        api_key (str): API key for authentication.
        server_url (str): URL of the Daytona server.
        target (str): Default target location for workspaces.
        workspace_api (WorkspaceApi): API client for workspace operations.
        toolbox_api (ToolboxApi): API client for toolbox operations.

    Example:
        Using environment variables:
        ```python
        daytona = Daytona()  # Uses DAYTONA_API_KEY, DAYTONA_SERVER_URL
        ```

        Using explicit configuration:
        ```python
        config = DaytonaConfig(
            api_key="your-api-key",
            server_url="https://your-server.com",
            target="us"
        )
        daytona = Daytona(config)
        ```
    """

    def __init__(self, config: Optional[DaytonaConfig] = None):
        """Initialize Daytona instance with optional configuration.

        If no config is provided, reads from environment variables:
        - `DAYTONA_API_KEY`: Required API key for authentication
        - `DAYTONA_SERVER_URL`: Required server URL
        - `DAYTONA_TARGET`: Optional target environment (defaults to "local")

        Args:
            config (Optional[DaytonaConfig]): Object containing api_key, server_url, and target.

        Raises:
            ValueError: If API key or Server URL is not provided either through config
                or environment variables.

        Example:
            ```python
            from daytona_sdk import Daytona, DaytonaConfig
            # Using environment variables
            daytona1 = Daytona()
            # Using explicit configuration
            config = DaytonaConfig(
                api_key="your-api-key",
                server_url="https://your-server.com",
                target="us"
            )
            daytona2 = Daytona(config)
            ```
        """
        if config is None:
            # Initialize env - it automatically reads from .env and .env.local
            env = Env()
            env.read_env()  # reads .env
            # reads .env.local and overrides values
            env.read_env(".env.local", override=True)

            self.api_key = env.str("DAYTONA_API_KEY")
            self.server_url = env.str("DAYTONA_SERVER_URL")
            self.target = env.str("DAYTONA_TARGET", "local")
        else:
            self.api_key = config.api_key
            self.server_url = config.server_url
            self.target = config.target

        if not self.api_key:
            raise ValueError("API key is required")

        if not self.server_url:
            raise ValueError("Server URL is required")

        # Create API configuration without api_key
        configuration = Configuration(host=self.server_url)
        api_client = ApiClient(configuration)
        api_client.default_headers["Authorization"] = f"Bearer {self.api_key}"

        # Initialize API clients with the api_client instance
        self.workspace_api = WorkspaceApi(api_client)
        self.toolbox_api = ToolboxApi(api_client)

    def create(self, params: Optional[CreateWorkspaceParams] = None) -> Workspace:
        """Creates a new workspace and waits for it to start.

        Args:
            params (Optional[CreateWorkspaceParams]): Parameters for workspace creation. If not provided,
                   defaults to Python language.

        Returns:
            Workspace: The created workspace instance.

        Raises:
            ValueError: If timeout or auto_stop_interval is negative.
            Exception: If workspace creation fails.

        Example:
            Create a default Python workspace:
            ```python
            workspace = daytona.create()
            ```

            Create a custom workspace:
            ```python
            params = CreateWorkspaceParams(
                language="python",
                name="my-workspace",
                image="debian:12.9",
                env_vars={"DEBUG": "true"},
                resources=WorkspaceResources(cpu=2, memory=4096),
                timeout=300,
                auto_stop_interval=0
            )
            workspace = daytona.create(params)
            ```
        """
        # If no params provided, create default params for Python
        if params is None:
            params = CreateWorkspaceParams(language="python")

        workspace_id = params.id if params.id else f"sandbox-{str(uuid.uuid4())[:8]}"
        code_toolbox = self._get_code_toolbox(params)

        try:
            if params.timeout and params.timeout < 0:
                raise ValueError("Timeout must be a non-negative number")

            if params.auto_stop_interval is not None and params.auto_stop_interval < 0:
                raise ValueError(
                    "auto_stop_interval must be a non-negative integer")

            # Create workspace using dictionary
            workspace_data = CreateWorkspace(
                id=workspace_id,
                name=params.name if params.name else workspace_id,
                image=params.image,
                user=params.os_user if params.os_user else "daytona",
                env=params.env_vars if params.env_vars else {},
                labels=params.labels,
                public=params.public,
                target=params.target if params.target else self.target,
                auto_stop_interval=params.auto_stop_interval
            )

            if params.resources:
                workspace_data.cpu = params.resources.cpu
                workspace_data.memory = params.resources.memory
                workspace_data.disk = params.resources.disk
                workspace_data.gpu = params.resources.gpu

            response = self.workspace_api.create_workspace(
                create_workspace=workspace_data)
            workspace = Workspace(
                workspace_id,
                response,
                self.workspace_api,
                self.toolbox_api,
                code_toolbox
            )

            # Wait for workspace to start
            try:
                workspace.wait_for_workspace_start(params.timeout)
            finally:
                # If not Daytona SaaS, we don't need to handle pulling image state
                pass

            return workspace

        except Exception as e:
            try:
                self.workspace_api.remove_workspace(workspace_id=workspace_id)
            except:
                pass
            raise Exception(f"Failed to create workspace: {str(e)}") from e

    def _get_code_toolbox(self, params: Optional[CreateWorkspaceParams] = None):
        """Helper method to get the appropriate code toolbox based on language.

        Args:
            params (Optional[CreateWorkspaceParams]): Workspace parameters. If not provided, defaults to Python toolbox.

        Returns:
            The appropriate code toolbox instance for the specified language.

        Raises:
            ValueError: If an unsupported language is specified.
        """
        if not params:
            return WorkspacePythonCodeToolbox()

        match params.language:
            case "javascript" | "typescript":
                return WorkspaceTsCodeToolbox()
            case "python":
                return WorkspacePythonCodeToolbox()
            case _:
                raise ValueError(f"Unsupported language: {params.language}")

    def remove(self, workspace: Workspace) -> None:
        """Removes a workspace.

        Args:
            workspace (Workspace): The workspace instance to remove.

        Example:
            ```python
            workspace = daytona.create()
            # ... use workspace ...
            daytona.remove(workspace)  # Clean up when done
            ```
        """
        return self.workspace_api.delete_workspace(workspace_id=workspace.id, force=True)

    def get_current_workspace(self, workspace_id: str) -> Workspace:
        """Get a workspace by its ID.

        Args:
            workspace_id (str): The ID of the workspace to retrieve.

        Returns:
            Workspace: The workspace instance.

        Raises:
            ValueError: If workspace_id is not provided.

        Example:
            ```python
            workspace = daytona.get_current_workspace("my-workspace-id")
            print(workspace.status)
            ```
        """
        if not workspace_id:
            raise ValueError("workspace_id is required")

        # Get the workspace instance
        workspace_instance = self.workspace_api.get_workspace(
            workspace_id=workspace_id)

        # Create and return workspace with Python code toolbox as default
        code_toolbox = WorkspacePythonCodeToolbox()
        return Workspace(
            workspace_id,
            workspace_instance,
            self.workspace_api,
            self.toolbox_api,
            code_toolbox
        )

    def list(self) -> List[Workspace]:
        """List all workspaces.

        Returns:
            List[Workspace]: List of all available workspace instances.

        Example:
            ```python
            workspaces = daytona.list()
            for workspace in workspaces:
                print(f"{workspace.id}: {workspace.status}")
            ```
        """
        workspaces = self.workspace_api.list_workspaces()
        return [
            Workspace(
                workspace.id,
                workspace,
                self.workspace_api,
                self.toolbox_api,
                self._get_code_toolbox(
                    CreateWorkspaceParams(
                        language=self._validate_language_label(
                            workspace.labels.get("code-toolbox-language"))
                    )
                )
            )
            for workspace in workspaces
        ]

    def _validate_language_label(self, language: Optional[str]) -> CodeLanguage:
        """Validate and normalize the language label.

        Args:
            language (Optional[str]): The language label to validate.

        Returns:
            CodeLanguage: The validated language, defaults to "python" if None

        Raises:
            ValueError: If the language is not supported.
        """
        if not language:
            return "python"

        if language not in ["python", "javascript", "typescript"]:
            raise ValueError(f"Invalid code-toolbox-language: {language}")

        return language  # type: ignore

    def start(self, workspace: Workspace, timeout: Optional[float] = None) -> None:
        """Starts a workspace and waits for it to be ready.

        Args:
            workspace (Workspace): The workspace to start.
            timeout (Optional[float]): Optional timeout in seconds to wait for the workspace to start. If set to 0, it will wait indefinitely.

        Example:
            ```python
            workspace = daytona.get_current_workspace("my-workspace-id")
            daytona.start(workspace, timeout=40)  # Wait up to 40 seconds
            ```
        """
        workspace.start(timeout)
        workspace.wait_for_workspace_start()

    def stop(self, workspace: Workspace) -> None:
        """Stops a workspace and waits for it to be stopped.

        Args:
            workspace (Workspace): The workspace to stop.

        Example:
            ```python
            workspace = daytona.get_current_workspace("my-workspace-id")
            daytona.stop(workspace)
            ```
        """
        workspace.stop()
        workspace.wait_for_workspace_stop()


# Export these at module level
__all__ = [
    "Daytona",
    "DaytonaConfig",
    "CreateWorkspaceParams",
    "CodeLanguage",
    "Workspace",
    "SessionExecuteRequest",
    "SessionExecuteResponse"
]
